### DISCUSSION TITLE
Welcome to btp-setup-automator Discussions!

### DISCUSSION THREAD

**Discussion by rui1610** [SAP-archive/btp-setup-automator#20] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by aliedl0424** [comment]
Greetings!  Alex Liedl from Cargill.  We are just getting started on our BTP journey and my goal is to embed automation from day 1.  I came across the btp setup automator and am intrigued.  Still havent gone through all the documentation or dig through all the included scripts but any thoughts on how we could tailor or create scripts for basic BTP operational activities such as creating a subaccount, managing users/role collections, trust settings, setting entitlements, subscribing to services, etc.  My goal is to have all of this documented in a state file somewhere similar to how we would manage infra with a tool such as terraform.  Thoughts?

**Reply by yevgentrukhin** [reply]
Hi Alex, the tool can do exactly what you described and the state is kept in usecase.json files. You would need to study a bit sample use case files to see how you can tailor them towards your usecase, check this one, for example: https://github.com/SAP-samples/btp-setup-automator/blob/main/usecases/released/extendS4_with_devops.json 
Best Regards, Yevgen

**Reply by rui1610** [reply]
Checkout the documentation providing you a step-by-step explanation:

https://github.com/SAP-samples/btp-setup-automator/tree/main/docs/README.md

**Comment by rui1610** [comment]
Hey @aliedl0424 , where you successful in getting started with the script? Interested to understand how you are using it and whether it is useful to you at all.

**Comment by aliedl0424** [comment]
Hey thanks for reaching out!  So I took a look at the scripts included.  I
see the potential.  That said, I see it not quite meeting my requirements.
Ideally I am looking for something like terraform that would be able to not
only deploy resources but also manage the state (i.e. check to see if tf
script has resources already provisioned in the tf.state file).  At this
point in our journey, i am not necessarily looking for a means to deploy
resources for full stack applications but rather purely on the BTP platform
operations (i.e. setting up sub accounts, enabling principle propagation,
assigning users and authorizations, setting up destinations, etc).  I
believe that the scripts could do this but i also see that there would be
quite some python reverse engineering required.  That said, if you are able
to provide a really simple example of one of the use cases mentioned above,
it may help me in my thought process.  I also heard that you may be working
on terraform support and may have some preliminary successes?  Would be
really interested in learning more/partnering.

**Reply by rui1610** [reply]
Did you have a look into the released usecase files? The documentation should help you as well to get up-to-speed on the btp-setup-automator.

**Comment by dachtera** [comment]
Hello all,

My team is working on building out our BTP automation for creation of sub-accounts and we're running into a few issues.

1) There is a discrepancy between the entitlements and options available within the UI, and those available when creating your JSON file (Such as Cloud Foundry, Launchpad to name a few of our core services)

2) Difficulty adding multiple plans in the JSON file, when it's possible within the UI.

Example below of a service available within the UI, but not  the automator:
ERROR      : service >saplaunchpad< with plan >standard< in region >eu10< IS NOT AVAILABLE

Our end goal is to have an end-to-end process to add in all of the entitlements/options etc for a sub-account without manual updates after the fact.   

Can I connect with someone with deeper experience in the scripting and learn more how/why we're running into discrepancies and how to overcome them?

Thank You.

**Reply by rui1610** [reply]
This was an issue we've detected yesterday morning, too, due to some recent changes in the creation of some libs. This was fixed and should be available since yesterday early afternoon. Simply fetch the newest docker image and it should work again.
In case you find such issues, the best way to address them is by adding an issue in this GitHub repo.
Hope this helped.

Best,
Rui

**Reply by dachtera** [reply]
Hi Rui- This issue still persists this morning, and I cannot proceed- 
ERROR      : service >saplaunchpad< with plan >standard< in region >eu10< IS NOT AVAILABLE

These are the three entitlements I'm trying to add in right now, and I also have an error about the 'standard' plan for CF is not available, only free, yet when I configure manually, I have the standard option.

```
{
            "name": "sapappstudio",
            "plan": "standard-edition",
            "category": "APPLICATION",
            "entitleonly": true
        },
        {
            "name": "cloudfoundry",
            "plan": "standard",
            "category": "ENVIRONMENT",
            "entitleonly": true
        },
        {
            "name": "saplaunchpad",
            "plan": "standard",
            "category": "APPLICATION",
            "entitleonly": true
        }

```
Another issue I encountered this morning, trying to add CloudFoundry:

```
[2022-10-13 15:35:28] INFO       : Check for available environment landscapes in subaccount >07f9e26d-ed9d-44e8-bcb4-8d19c9da3aaa< and region >eu10< (try 5 - trying again in 3s)
[2022-10-13 15:35:28] RUN COMMAND: btp --format json list accounts/available-environment --subaccount '07f9e26d-ed9d-44e8-bcb4-8d19c9da3aaa'
[2022-10-13 15:35:34] ERROR      : Traceback (most recent call last):

NoneType: None
[2022-10-13 15:35:34] ERROR      :   File "/home/user/./btpsa", line 32, in <module>
    myBtpUseCase.initialize_environments()

NoneType: None
[2022-10-13 15:35:34] ERROR      :   File "/home/user/libs/python/btp_cli.py", line 287, in initialize_environments
    self.create_environments()

NoneType: None
[2022-10-13 15:35:34] ERROR      :   File "/home/user/libs/python/btp_cli.py", line 323, in create_environments
    envLandscape = selectEnvironmentLandscape(

NoneType: None
[2022-10-13 15:35:34] ERROR      :   File "/home/user/libs/python/btp_cli.py", line 1488, in selectEnvironmentLandscape
    environment["name"] + "< for >" + region + "<")

NoneType: None
[2022-10-13 15:35:34] ERROR      : TypeError: 'BTPSERVICE' object is not subscriptable

NoneType: None
[2022-10-13 15:35:34] ERROR      : 'BTPSERVICE' object is not subscriptable
NoneType: None
```

Unsure if they are bugs, or simply my lack of knowing how everything is keyed in, even when looking at docs.  I can start logging issues thought!

My teams end goal is an end-to-end automation for starting with the global account creation, and then the sub account with entitlements, users, permissions (which are also not applying), CF org members, etc.   It may be either A, I'm missing knowledge, or B, it's not possible.  We are migrating all of our former COLO customers onto BTP, and the manual effort is staggering..

**Reply by rui1610** [reply]
The name of the service is wrong. It is case sensitive and is called SAPLaunchpad

**Reply by rui1610** [reply]
You can simply remove the entry for cloudfoundry. Should work as well.

**Reply by dachtera** [reply]
Interesting on the cases, as I'm using VS Code with type-ahead enabled.  Seems to be an issue there.   Actually, earlier in the week when I was manually keying everything SAPLaunchpad was throwing an error about now being available but processes now ( keeps looping this now for 10 minutes and counting)
```
[2022-10-13 15:48:36] INFO       : check status of app subscriptions
[2022-10-13 15:48:36] RUN COMMAND: btp --format json list accounts/subscription --subaccount '07f9e26d-ed9d-44e8-bcb4-8d19c9da3aaa'
```

CloudFoundry is a requirement for our team, as well as creating the space and assigning the org members.  If the Automator is not capable of completing these tasks, it's not terribly useful to our team.

Would it be wise of me to setup a meeting with someone to help determine what is my mistake and what is truly a bug/issue?  I don't want to clutter up things by submitting a bunch of issues, if it's a quick fix to my JSON.

**Reply by rkapoz-bgp** [reply]
> You can simply remove the entry for cloudfoundry. Should work as well.

Hi rui1610,
I'm starting using and testing this automation tool; I noticed a strange thing about services that the automation provision: the automation provision CF even is not explicitily referenced in the list of services to be provisioned. May be the case other defined services that depend on CF will cause CF to be provisioned anyway?

**Reply by lechnerc77** [reply]
@dachtera: I will try to split your questions into dedicated units:

- *"SAPLaunchpad"-topic*: The naming/type ahead support was a bit shaky due to a switch in the internal fetching of the information. This was fixed and should be stable now.
- *Creation of Cloud Foundry spaces*: You can create Cloud Foundry spaces, indeed this is the default behavior. As you can see in the `parameters.json` you can specify e.g. the org id, org name and the space name that should be created and/or used for the service creation.
- *"Infinite loop in entitlements"-topic*: The entitlement of application contained a bug which caused this behavior. It was fixed via PR #258. A new Docker image containing the fix is available. 
- *Setup*: The desired target setup you described in this discussion works and is described in issue #255 (including sample files). The created Cloud Foundry environment contains only the person specified in the `email` field of the `parameters.json` as Org Manager. 
- *Adding members to Cloud Foundry in different roles*: If you want to add additional members in different roles you need to specify them in the `parameters.json` via `"usergroups"`. You can define several user groups. These groups must then be referenced in the `usecase.json` file. You find a sample in the file [cf_cups.json](https://github.com/SAP-samples/btp-setup-automator/blob/main/usecases/released/cf_cups.json) under `usecases/released`. 
- *Connection Service and Environment*: The connection between services and environments is defined by the optional field `"targetenvironment"` that you can define per service. The default value is Cloud Foundry, but you can also create services in Kyma (which comes with the prerequisite that the `kubeconfig.json` is available in the container) and services for the environment `"others"` (as in the SAP BTP Cockpit) which will then have no binding to any environment. The same is true for service bindings. 

Hope that helps

P.S. in order to keep the discussion a bit better manageable, please open dedicated question in the discussions area via starting a new discussion and selecting the category "Q&A"

**Comment by aliedl0424** [comment]
I would also love to be involved if there is a session on this.

**Comment by dachtera** [comment]
The enablement for cloud foundry in the automation is a requirement.  Simply removing it will not meet the requirements to have it added during automation.  Is there a second round of activations to take place after the initial sub account is created?  I may really need to get on a call with someone to show our current environment and requirement for automation as I need to answer to my leadership on this process.


Thank You,

Tim Dachtera | CTE Consultant III | SAP Concur
***@***.******@***.***> | 952-697-2942 / 507-298-5355 | MN60 – Minnesota, US

On Oct 15, 2022, at 3:58 AM, Roberto Capozzo ***@***.***> wrote:

﻿

You can simply remove the entry for cloudfoundry. Should work as well.

Hi rui1610,
I'm starting using and testing this automation tool; I noticed a strange thing about services that the automation provision: the automation provision CF even is not explicitily referenced in the list of services to be provisioned. May be the case other defined services that depend on CF will cause CF to be provisioned anyway?

**Reply by lechnerc77** [reply]
See my answers above <https://github.com/SAP-samples/btp-setup-automator/discussions/20#discussioncomment-3885852>
